### ISSUE TITLE
[QUESTION] Any way to do query with group by rounded date times via linq?

### ISSUE THREAD

**Issue by grofit** [litedb-org/LiteDB#2270]
I have a scenario where I often have to show lots of graphs which group data at per hour/minute level, so if I were doing it in memory I would get the data and do something like:

```csharp
var chartData = unfilteredDataInMemory
   .GroupBy(x => x.EventDate.RoundToNearestMinute()) // makes all events be grouped per minute (or hour/day etc)
   .Select(x => new KeyedValue(x.Key, x.Sum(y => y.EventValue))); // gets the total of all event values within that minute
```

So there could be hundreds of events per minute so having to pull this all out in memory and do all the computations there can be REALLY slow over larger date periods, however at the DB level this should be fine, however I couldnt find a way to express this via the v5 LiteDb functionality.

So is there a way to GroupBy the date but transform the date in a way that lets it key everything within the same minute etc (be it via string formatting or some built in method etc.

Any help would be great, thanks.

**Comment by mbdavid** [comment]
Hi @grofit,

In this scenario, it's recommended created an adicional property in your class, like "EventDateGroup" with your RoundToNearestMinute truncate function applied (can return a long - ticks). In this case, you will store this group data in a stored value (not an expression). If you have a large data volume, create an index in this field for better results

After that, you can linq your expression using this adicional property.

Best regards

**Comment by grofit** [comment]
Hey, thanks for getting back to me so quickly, unfortunatly that rounding amount is a dynamic value provided at query time from a user front end, i.e they may want minute level granularity for more focused checks but hour granularity for overviews of things.

I did find I could somewhat sidestep this issue to some extent by doing a substring to only take a certain amount of chars i.e 16 for minute level, but I just wanted to know if there was any `DATEFORMAT(date, "YYYY-MM-DD")` style method or something.

**Comment by grofit** [comment]
Incase anyone else finds this, I messed about a bit in the LiteDB Studio and found I could do what I needed to via some funky querying:

```sql
SELECT {
    Key: @key, -- Gets the date string
    MessageCount: COUNT(FILTER(*.EventType => @ = "chat-message")), -- gets the count of events of type "chat-message"
    UsersJoined: COUNT(FILTER(*.EventType => @ = "user-joined")), -- gets the count of events of type "user-joined"
    UsersLeft: COUNT(FILTER(*.EventType => @ = "user-left")),
    ViewerCount: AVG(
        MAP(FILTER(MAP(* => {type: @.EventType, value: @.EventValue}) => @.type = "viewer-count") => @.value)
    ) -- this is crazy it maps out the {type,value} takes only types of "viewer-count" then maps out JUST the value and averages it
    
}
FROM analytics_events
WHERE $.EventDateTime < "2023-01-09" -- This where could be anything
GROUP BY SUBSTRING($.EventDateTime, 0, 16) -- This cuts off anything past the minutes (assuming ISO 8601 format)
```

I think everything I want to do is possible within LiteDB it just takes some creativity, but that takes me half a second to query over 25k(ish) rows.

**Comment by JKamsker** [comment]
Yes: `FORMAT(date, format)` formats a date inside the query, and the format can be a parameter, so the user can pick the granularity at query time:

```csharp
var fmt = "yyyy-MM-dd HH:mm";   // or "yyyy-MM-dd HH", "yyyy-MM-dd", ...

var buckets = col.Query()
    .Where(x => x.EventDate >= from)
    .GroupBy(BsonExpression.Create("FORMAT($.EventDate, @0)", fmt))
    .Select("{ key: @key, total: SUM(*.EventValue) }")
    .ToList();
```

or in SQL:

```sql
SELECT { key: @key, total: SUM(*.EventValue) }
FROM analytics_events
WHERE $.EventDate >= @1
GROUP BY FORMAT($.EventDate, @0)
```

This needs `EventDate` stored as a date (a `DateTime` property). The buckets are in local time and match an in-memory `GroupBy(x => x.EventDate.ToString(fmt))`. It works on 5.0.21.

The 6.x prereleases also add a typed LINQ `GroupBy`:

```csharp
col.Query().GroupBy(x => x.EventDate.ToString(fmt))
   .Select(g => new KeyedValue { Key = g.Key, Total = g.Sum(y => y.EventValue) })
   .ToList();
```

Grouping still reads every document the `WHERE` matches (aggregates don't use indexes, see #1539), so keep the date range tight.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
